### PR TITLE
ARROW-1783: [Python] Provide a "component" dict representation of a serialized Python object with minimal allocation

### PR DESCRIPTION
**Pull request by wesm**
For systems (like Dask) that prefer to handle their own framed buffer transport, this provides a list of memoryview-compatible objects with minimal copying / allocation from the input data structure, which can similarly be zero-copy reconstructed to the original object.

To motivate the use case, consider a dict of ndarrays:

```
data = {i: np.random.randn(1000, 1000) for i in range(50)}
```

Here, we have:

```
>>> %timeit serialized = pa.serialize(data)
52.7 µs ± 1.01 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

This is about 400MB of data. Some systems may not want to double memory by assembling this into a single large buffer, like with the `to_buffer` method:

```
>>> written = serialized.to_buffer()
>>> written.size
400015456
```

We provide a `to_components` method which contains a dict with a `'data'` field containing a list of `pyarrow.Buffer` objects. This can be converted back to the original Python object using `pyarrow.deserialize_components`:

```
>>> %timeit components = serialized.to_components()
73.8 µs ± 812 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

>>> list(components.keys())
['num_buffers', 'data', 'num_tensors']

>>> len(components['data'])
101

>>> type(components['data'][0])
pyarrow.lib.Buffer
```

and

```
>>> %timeit recons = pa.deserialize_components(components)
93.6 µs ± 260 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

The reason there are 101 data components (1 + 2 * 50) is that:

* 1 buffer for the serialized Union stream representing the object
* 2 buffers for each of the tensors: 1 for the metadata and 1 for the tensor body. The body is separate so that this is zero-copy from the input

Next step after this is ARROW-1784 which is to transport a pandas.DataFrame using this mechanism

cc @pitrou @jcrist @mrocklin

### CODE REVIEW

File: cpp/src/arrow/ipc/message.h [hunk 1]
```diff
@@ -144,33 +144,20 @@ class ARROW_EXPORT MessageReader {
...
   /// \brief Read next Message from the interface
   ///
   /// \param[out] message an arrow::ipc::Message instance
   /// \return Status
   virtual Status ReadNextMessage(std::unique_ptr<Message>* message) = 0;
 };
 
-/// \brief Implementation of MessageReader that reads from InputStream
-/// \since 0.5.0
-class ARROW_EXPORT InputStreamMessageReader : public MessageReader {
```
Comment (wesm) [inline cpp/src/arrow/ipc/message.h]: It was never necessary to export this class

**Comment by mrocklin** [comment]
At first glance the `to_components`/`deserialize_components` structure seems good to me.  This is definitely something that we can work with on the Dask side.  Is it possible to easily turn each of the elements of `to_components()['data']` into a `memoryview` without significant cost?

**Comment by mrocklin** [comment]
```python
>>> list(components.keys())
['num_buffers', 'data', 'num_tensors']
```

I'm curious what the metadata here is supposed to mean and how it relates to `data`.  Just curious though.

**Comment by wesm** [comment]
Yeah, you just call `memoryview` on the objects in `to_components()['data']` -- the support the the buffer/memoryview protocol.

The serialized payload consists of an Arrow data structure describing the whole object, and the ndarrays / buffers are sent as "sidecars". So if `num_buffers` and `num_tensors` are both 0, then there will only be one buffer in `data`. In general there are `1 + num_buffers + 2 * num_tensors`

**Comment by wesm** [comment]
I could possibly add an argument to `to_components` that returns buffers as memoryviews, if that would help. Might be nice to encapsulate this detail somewhat (though Dask must have knowledge of the additional metadata anyhow)

**Comment by mrocklin** [comment]
From a Dask perspective I'm more than happy to handle the memoryview conversion.

File: cpp/src/arrow/python/python_to_arrow.cc [hunk 2]
```diff
@@ -710,27 +712,88 @@ Status SerializeObject(PyObject* context, PyObject* sequence, SerializedPyObject
...
+Status SerializedPyObject::GetComponents(MemoryPool* memory_pool, PyObject** out) {
+  PyAcquireGIL py_gil;
+
+  ScopedRef result(PyDict_New());
+  PyObject* buffers = PyList_New(0);
+
+  // TODO(wesm): Not sure how pedantic we need to be about checking the return
+  // values of these functions. There are other places where we do not check
+  // PyDict_SetItem/SetItemString return value, but these failures would be
+  // quite esoteric
```
Comment (pitrou) [inline cpp/src/arrow/python/python_to_arrow.cc]: The main failure mode would be MemoryError when growing the dict to make place for the new key.

File: cpp/src/arrow/python/python_to_arrow.cc [hunk 3]
```diff
@@ -710,27 +712,88 @@ Status SerializeObject(PyObject* context, PyObject* sequence, SerializedPyObject
...
+  PyDict_SetItemString(result.get(), "data", buffers);
+  RETURN_IF_PYERROR();
+
+  Py_DECREF(buffers);
+
+  auto PushBuffer = [&buffers](const std::shared_ptr<Buffer>& buffer) {
+    PyObject* wrapped_buffer = wrap_buffer(buffer);
+    RETURN_IF_PYERROR();
+    if (PyList_Append(buffers, wrapped_buffer) < 0) {
+      RETURN_IF_PYERROR();
```
Comment (pitrou) [inline cpp/src/arrow/python/python_to_arrow.cc]: You probably need `Py_DECREF(wrapper_buffer)` here as well.

File: cpp/src/arrow/ipc/message.h [hunk 4]
```diff
@@ -144,33 +144,20 @@ class ARROW_EXPORT MessageReader {
  public:
   virtual ~MessageReader() = default;
 
+  /// \brief Create MessageReader that reads from InputStream
+  static std::unique_ptr<MessageReader> Open(io::InputStream* stream);
+
+  /// \brief Create MessageReader that reads from owned InputStream
+  static std::unique_ptr<MessageReader> Open(
```
Comment (pitrou) [inline cpp/src/arrow/ipc/message.h]: For the record, is there a rationale or convention for the use of unique_ptr vs shared_ptr here? :-)

File: cpp/src/arrow/python/python_to_arrow.cc (hunk 3 above)
Comment (wesm) [reply cpp/src/arrow/python/python_to_arrow.cc]: done

File: cpp/src/arrow/ipc/message.h (hunk 4 above)
Comment (wesm) [reply cpp/src/arrow/ipc/message.h]: I try to only use shared_ptr when there is some reasonable expectation that shared ownership may be frequently needed (of course one can always transfer the pointer to a shared_ptr if needed). There are some other places in the library where `shared_ptr` is returned (or an out-variable) that would be better as `unique_ptr`

**Comment by wesm** [comment]
+1

**Pull request closed without merging**